### PR TITLE
Fix --override and --patch flags in kubectl

### LINKED ISSUE #4008: kubectl update --patch does not work as expected
I create a service
```
{
  "kind": "Service",
  "id": "web",
  "apiVersion": "v1beta1",
  "namespace": "default",
  "port": 9293,
  "protocol": "TCP",
  "selector": {
    "version": "v1"
  },
  "containerPort": 9293
}
```
update it with
```
# kubectl update se web --patch='{"selector":{"version":"v2"},"apiVersion":"v1beta1"}'
# kubectl describe se web
Name:           web
Labels:         <none>
Selector:       version=v1
Port:           9293
No events.
```

I would expect that the selector would have been updated but it's not. This is because of mergo.Merge which says in the doc that 'Merge sets fields' values in dst from src if they have a zero value of their type.' but `.selector.version` did not have a zero value before the merge. Currently patch can only add missing fields to api objects. Was this the intended behavior? Patch might be more useful if it could override existing fields.There is an open PR imdario/mergo#8 that would add the ability to override fields.

Heres a failing test: https://github.com/mikedanese/kubernetes/commit/a800ae6db81934bb66c5260452e4a214bad137ec

### PR DESCRIPTION
**Pull request by mikedanese**
Implements RFC7386 Json Merge Patch and uses it to fix misbehaving pkg/kubectl/cmd/uitil.Merge(). This should fix unexpected behavior of --override and --patch. The versioned patch is applied directly to versioned json of the resource so it should work for applying default values. Ultimately it might be nice to move the --patch update of a resource into the apiserver accessed by HTTP PATCH Method to the resources endpoint with `Content-Type: "application/merge-patch+json"`.

Fixes #4008

/cc @a-robinson @bgrant0607

### CODE REVIEW

**Comment by smarterclayton** [comment]
Can you put merge in a subpackage of util?  Util is growing unboundedly

**Comment by mikedanese** [comment]
Done.

**Comment by smarterclayton** [comment]
LGTM

File: pkg/kubectl/cmd/update.go [hunk 1]
```diff
@@ -116,12 +116,13 @@ func updateWithPatch(cmd *cobra.Command, args []string, f *Factory, patch string
...
-	cmdutil.Merge(obj, patch, mapping.Kind)
+	patchedObj, err := cmdutil.Merge(obj, patch, mapping.Kind)
+	checkErr(err)
 
-	data, err := helper.Codec.Encode(obj)
+	data, err := helper.Codec.Encode(patchedObj)
 	checkErr(err)
 
-	obj, err = helper.Update(namespace, name, true, data)
+	_, err = helper.Update(namespace, name, true, data)
```
Comment (a-robinson) [inline pkg/kubectl/cmd/update.go]: Does this actually fix the handling of default-valued field updates when the object gets re-marshalled?

Comment (mikedanese) [inline pkg/kubectl/cmd/update.go]: This particular line does not but the PR does. Check out merge_test.go. Just to be sure i've added some test to helper_test.go as well. See
https://github.com/GoogleCloudPlatform/kubernetes/pull/4448/files#diff-c8e282cc368fe38828935844bb3bd00aR113
which delete's a field and
https://github.com/GoogleCloudPlatform/kubernetes/pull/4448/files#diff-c8e282cc368fe38828935844bb3bd00aR133
which updates an int field to zero. default-value updates are fixed by applying patch json directly to versioned type json.

Comment (mikedanese) [inline pkg/kubectl/cmd/update.go]: I think I see why you put the comment on this line. On the trip back to the server are the default fields wiped out? I believe they still are if the field is annotated with omitempty. For that we would have to eliminate round trip in kubectl.

Comment (mikedanese) [inline pkg/kubectl/cmd/update.go]: Or implement HTTP PATCH on our resource endpoints.

Comment (a-robinson) [inline pkg/kubectl/cmd/update.go]: Thanks for clarifying, I just wanted to double check that this wasn't supposed to be fixing everything we talked about in the comments of #4008, just the original issue.

I got interrupted last night when going through this, I'll check out the rest now.

**Comment by a-robinson** [comment]
LGTM, thanks Mike!

**Comment by mikedanese** [comment]
I believe this does actually work for default fields even with omitempty annotation since it looks like helper.Update() overwrites the object so empty default fields would be overwritten even if they don't exist in the json that kubectl sends to the apiserver. I'm going to test this though. Good question, sorry for the noise.

**Comment by ghodss** [comment]
@mikedanese Could you add the PATCH verb to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md#verbs-on-resources along with an explanation of how defaults, empty fields, etc. are handled?

**Comment by mikedanese** [comment]
@ghodss Currently PATCH is not implemented in the apiserver only in kubectl. I was thinking about knocking that out in the next couple of days if this feature is wanted. I'll open an issue soon to discuss this. Currently merge happens in kubectl and follows the convention defined in [RFC7368](https://tools.ietf.org/html/rfc7386). I'd document this somewhere but I'm not sure where the best place is.

**Comment by mikedanese** [comment]
@ghodss if you are interested, I've opened #4578 to discuss PATCH.

**Pull request merged**